sdk/src/test/scala/com/horizen/consensus/ConsensusDataProviderTest.scala

```diff
@@ -0,0 +1,272 @@
+}
+
+class ConsensusDataProviderTest extends CompanionsFixture{
+  val generator: SidechainBlockFixture = new SidechainBlockFixture {}
+  val dummyWithdrawalEpochInfo = utils.WithdrawalEpochInfo(0, 0)
+
+
+  @Test
+  def test(): Unit = {
+    val slotsInEpoch = 10
+    val slotsPresentation: List[List[Int]] = List(
+//   1 -- block in slot is present; 0 -- no block for slot
+//   slots 1  2  3  4  5  6  7  8  9  10
+      List(1, 1, 1, 1, 1, 1, 1, 1, 1, 1), //2 epoch
+      List(1, 1, 0, 0, 1, 1, 1, 1, 0, 0), //3 epoch
+      List(1, 1, 0, 0, 1, 1, 1, 1, 0, 0), //4 epoch
+      List(0, 0, 0, 0, 1, 1, 1, 1, 0, 0), //5 epoch
+      List(1, 0, 0, 0, 0, 0, 0, 0, 0, 1), //6 epoch
+      List(1, 0, 1, 0, 1, 0, 1, 0, 1, 0), //7 epoch
+      List(1, 0, 1, 0, 1, 0, 1, 0, 1, 0), //8 epoch
+      List(0, 1, 1, 0, 1, 0, 1, 1, 1, 1), //9 epoch
+      List(0, 1, 1, 0, 1, 0, 1, 1, 0, 1), //10 epoch
+
+    )
+    require(slotsPresentation.forall(_.size == slotsInEpoch))
+
+    val genesisBlockId = bytesToId(Utils.doubleSHA256Hash("genesis".getBytes()))
+    val genesisBlockTimestamp = 1000000
+    val networkParams = TestNetParams(
+      sidechainGenesisBlockId = genesisBlockId,
+      sidechainGenesisBlockTimestamp = genesisBlockTimestamp,
+      consensusSlotsInEpoch = slotsInEpoch,
+      consensusSecondsInSlot = 100
+    )
+
+    val tested1 = new TestedConsensusDataProvider(slotsPresentation, networkParams)
+    val blockIdAndInfosPerEpoch1 = tested1.blockIdAndInfosPerEpoch
+    val epochIds1 = tested1.epochIds
+    //Finished preparation
+
+    val consensusInfoForGenesisEpoch = tested1.getInfoForCheckingBlockInEpochNumber(1)
+
+    val blockIdAndInfoFromStartSecondEpoch = blockIdAndInfosPerEpoch1(1).head
+    val consensusInfoForStartSecondEpoch = tested1.getFullConsensusEpochInfoForBlock(blockIdAndInfoFromStartSecondEpoch)
+
+    val blockIdAndInfoFromEndSecondEpoch = blockIdAndInfosPerEpoch1(1).last
+    val consensusInfoForEndSecondEpoch = tested1.getFullConsensusEpochInfoForBlock(blockIdAndInfoFromEndSecondEpoch)
+
+    val consensusInfoForEndThirdEpoch = tested1.getInfoForCheckingBlockInEpochNumber(3)
+
+    val consensusInfoForEndFourthEpoch = tested1.getInfoForCheckingBlockInEpochNumber(4)
+
+    //Full consensus info is the same for any block in epoch
+    assertEquals(consensusInfoForStartSecondEpoch, consensusInfoForEndSecondEpoch)
+
+    //Stake and root hash is the same for second and genesis epoch
+    assertEquals(consensusInfoForEndSecondEpoch.stakeConsensusEpochInfo, consensusInfoForGenesisEpoch.stakeConsensusEpochInfo)
+    //and stake is as expected
+    assertEquals(1000, consensusInfoForEndSecondEpoch.stakeConsensusEpochInfo.totalStake)
+    //nd stake root hash as expected
+    assertTrue(epochIds1.head.getBytes.take(merkleTreeHashLen).sameElements(consensusInfoForGenesisEpoch.stakeConsensusEpochInfo.rootHash))
+    //but nonce is the same as well // Is it acceptable? //
```

Comment:
This should be discussed.
Previously nonce for genesis info for some "fixed" value. Nonce for second epoch was equal to mc ref hash in genesis block.
Now, nonce calculation algorithm is different and genesis block is a part of "quite" slots.
Maybe, we need to have a special case for second epoch nonce calculation and take genesis block vrf hash in consideration.